template-coq/gen-src/run_extractable.ml

```diff
@@ -0,0 +1,103 @@
+open Extractable
+open Plugin_core
+
+(* todo(gmm): determine what of these already exist. *)
+let to_constr (t : Ast0.term) : Constr.t =
+  failwith "to_constr"
+
+let of_constr (t : Constr.t) : Ast0.term =
+  failwith "of_constr"
+
+let to_string (cl : char list) : string =
+  failwith "to_string"
+
+let of_string (s : string) : char list =
+  failwith "of_string"
+
+let to_reduction_strategy (s : Common.reductionStrategy) =
+  failwith "to_reduction_strategy"
+
+let to_ident : char list -> Names.Id.t =
+  failwith "to_ident"
+
+let of_ident : Names.Id.t -> char list =
+  failwith "of_ident"
+
+let of_global_reference (t : global_reference) : BasicAst.global_reference =
+  failwith "of_global_reference"
+
+let to_qualid (c : char list) : Libnames.qualid =
+  Libnames.qualid_of_string (to_string c)
+
+let of_qualid (q : Libnames.qualid) : char list =
+  of_string (Libnames.string_of_qualid q)
+
+let of_kername : Names.KerName.t -> char list =
+  failwith "of_kername"
+
+let to_kername : char list -> Names.KerName.t =
+  failwith "of_kername"
+
+let of_mib : Plugin_core.mutual_inductive_body -> _ =
+  failwith "of_mib"
+
+let to_mie : _ -> Plugin_core.mutual_inductive_entry =
+  failwith "to_mie"
+
+let of_constant_entry : _ -> Ast0.constant_entry =
+  failwith "of_constant_entry"
+
+let rec interp_tm (t : 'a coq_TM) : 'a tm =
+  match t with
+  | Coq_tmReturn x -> tmReturn x
+  | Coq_tmBind (c, k) -> tmBind (interp_tm c) (fun x -> interp_tm (k x))
+  | Coq_tmPrint t -> Obj.magic (tmPrint (to_constr t))
+  | Coq_tmMsg msg -> Obj.magic (tmMsg (to_string msg))
+  | Coq_tmFail err -> tmFail (to_string err)
+  | Coq_tmEval (r,t) ->
+    tmMap (fun x -> Obj.magic (of_constr x))
+          (tmEval (to_reduction_strategy r) (to_constr t))
+  | Coq_tmDefinition (nm, typ, trm) ->
+    let typ =
+      match typ with
+        None -> None
+      | Some typ -> Some (to_constr typ)
+    in
+    tmMap (fun x -> Obj.magic (of_kername x))
+          (tmDefinition (to_ident nm) typ (to_constr trm))
```

Comment:
Are we assuming here that `to_constr` does unquoting as well?

Reply:
What does "unquoting" mean? I think `to_constr` is just one direction of the isomorphism, it shouldn't need to do anything intelligent.

Reply:
Ok. It's body would be very similar to the following? https://github.com/MetaCoq/metacoq/blob/d5088ac77736bb1576fbc815aed07c2dae3562dd/template-coq/src/denote.ml#L347

Reply:
It should be similar, except the type will be a little bit different.

Reply:
ok. I am working on it (`to_constr`).

Reply:
In #11, `denote_term` was sufficiently functorized so that we could have used its instantiation as `to_constr`. Somehow, those changes to `denote_term` were undone since then. My last [commit](https://github.com/MetaCoq/metacoq/pull/113/files/92ff62c93ce03d1c3f7a7432d8bc23b6904655ef) introduces a lot of code duplication and I think going back to the functorized version of `denote_term` may be the right approach.